File: atosim/org.palladiosimulator.blockchainsystems.threesim/src/main/java/org/palladiosimulator/blockchainsystems/threesim/metrics/calculators/AttackerRevenueShareCalculator.java
package org.palladiosimulator.blockchainsystems.threesim.metrics.calculators;

import org.palladiosimulator.blockchainsystems.threesim.metrics.AttackerRevenueShare;
import org.palladiosimulator.blockchainsystems.threesim.metrics.abstractions.AverageOutputMetric;
import org.palladiosimulator.blockchainsystems.threesim.metrics.abstractions.AverageOutputMetricImpl;
import java.util.List;
import java.util.stream.Collectors;

public class AttackerRevenueShareCalculator {
    private final double attackerRewards;
    private final double totalRewards;

    public AttackerRevenueShareCalculator(double attackerRewards, double totalRewards) {
        this.attackerRewards = attackerRewards;
        this.totalRewards = totalRewards;
    }

    public AttackerRevenueShare calculate() {
        double value = totalRewards == 0.0 ? 0.0 : (attackerRewards / totalRewards) * 100.0;
        return new AttackerRevenueShare(value);
    }

    public static AverageOutputMetric calculateAverage(List<AttackerRevenueShare> values) {
        double avg = values.stream().mapToDouble(AttackerRevenueShare::getValue).average().orElse(0.0);
        double variance = values.stream().mapToDouble(v -> (v.getValue() - avg) * (v.getValue() - avg)).average().orElse(0.0);
        double std = Math.sqrt(variance);
        Double cv = avg == 0.0 ? null : std / avg;
        return new AverageOutputMetricImpl(AttackerRevenueShare.NAME, avg, AttackerRevenueShare.UNIT, std, cv);
    }
}

File: atosim/org.palladiosimulator.blockchainsystems.threesim/src/main/kotlin/org/palladiosimulator/blockchainsystems/threesim/metrics/calculators/AttackerRevenueShareCalculator.kt
package org.palladiosimulator.blockchainsystems.threesim.metrics.calculators

import org.palladiosimulator.blockchainsystems.threesim.metrics.AttackerRevenueShare
import org.palladiosimulator.blockchainsystems.threesim.metrics.abstractions.AverageOutputMetric
import org.palladiosimulator.blockchainsystems.threesim.metrics.abstractions.AverageOutputMetricImpl
import kotlin.math.sqrt

class AttackerRevenueShareCalculator(
    private val attackerRewards: Double,
    private val totalRewards: Double
) {

    fun calculate(): AttackerRevenueShare {

        val value =
            if (totalRewards == 0.0) 0.0
            else (attackerRewards / totalRewards) * 100.0

        return AttackerRevenueShare(value)
    }

    companion object {

        fun calculateAverage(values: List<AttackerRevenueShare>): AverageOutputMetric {

            val avg = values.map { it.value }.average()

            val variance = values
                .map { (it.value - avg) * (it.value - avg) }
                .average()

            val std = sqrt(variance)

            val cv = if (avg == 0.0) null else std / avg

            return AverageOutputMetricImpl(
                name = AttackerRevenueShare.NAME,
                average = avg,
                unit = AttackerRevenueShare.UNIT,
                standardDeviation = std,
                coefficientOfVariation = cv
            )
        }
    }
}